cluster-operator/src/main/java/io/strimzi/operator/cluster/leaderelection/LeaderElectionManager.java

```
@@ -0,0 +1,123 @@
+        this.client = client;
+        this.config = config;
+        this.startLeadershipCallback = startLeadershipCallback;
+        this.stopLeadershipCallback = stopLeadershipCallback;
+        this.leadershipChangeCallback = leadershipChangeCallback;
+
+        this.managerThread = new Thread(this);
+    }
+
+    /**
+     * The run() method which just calls into the Fabric8 Kubernetes client API.
+     */
+    @Override
+    public void run() {
+        // TODO: Do we need handle interrupted here any exit?
+        client.leaderElector()
+                .withConfig(
+                        new LeaderElectionConfigBuilder()
+                                .withName(config.getLeaseName())
+                                .withLock(new LeaseLock(config.getNamespace(), config.getLeaseName(), config.getIdentity()))
+                                .withLeaseDuration(config.getLeaseDuration())
+                                .withRenewDeadline(config.getRenewDeadline())
+                                .withRetryPeriod(config.getRetryPeriod())
+                                .withLeaderCallbacks(new LeaderCallbacks(
+                                        () -> startLeadership(),
+                                        () -> stopLeadership(),
+                                        newLeader -> leadershipChange(newLeader)))
+                                .withReleaseOnCancel(true)
+                                .build())
+                .build().run();
+
+        // Existing the leader election thread => if we are leader, we will loose the leadership
+        if (isLeader)   {
+            stopLeadership();
+        }
+    }
+
+    /**
+     * Start method to start the election thread
+     */
+    public void start() {
+        managerThread.start();
+    }
+
+    /**
+     * Stop method to stop the election thread
+     */
+    public void stop()  {
+        managerThread.interrupt();
```

is there any `InterruptedException` thrown in the run that we should catch and handle gracefully instead of having it in the log? Just asking because this exception is not always thrown anyway, so asking if you noticed it by any chance.

---

The Fabric8 method catches the interrupts and does not throw them. So we cannot catch it, but it just finishes the thread basically. And when that happens it looses the leadership and exits.

---

Yes it's what I was wondering. So we are covered.